File: platforms/android/nimbus/src/main/java/com/salesforce/nimbus/Bridge.kt
//
// Copyright (c) 2019, Salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
//

package com.salesforce.nimbus

import android.annotation.SuppressLint
import android.webkit.JavascriptInterface
import android.webkit.WebView
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

@SuppressLint("SetJavaScriptEnabled", "JavascriptInterface")
class Bridge {

    companion object {
        private const val BRIDGE_NAME = "_nimbus"
    }

    private var bridgeWebView: WebView? = null
    private val binders = mutableListOf<Binder>()

    /**
     * Adds a [NimbusBinder] to the bridge.
     */
    fun add(vararg binder: Binder) {
        binders.addAll(binder)
    }

    /**
     * Attaches the bridge to the provided [webView], initializing each extension and loading the
     * [appUrl].
     */
    fun attach(webView: WebView) {
        bridgeWebView = webView
        if (!webView.settings.javaScriptEnabled) {
            webView.settings.javaScriptEnabled = true
        }
        webView.addJavascriptInterface(this, BRIDGE_NAME)
        initialize(webView, binders)
    }

    fun loadUrl(appUrl: String) {
        bridgeWebView?.loadUrl(appUrl)
    }

    /**
     * Detaches the bridge performing any necessary cleanup.
     */
    fun detach() {
        bridgeWebView?.let { webView ->
            webView.removeJavascriptInterface(BRIDGE_NAME)
            cleanup(webView, binders)
        }
        binders.clear()
        bridgeWebView = null
    }

    private fun invokeInternal(
        identifierSegments: Array<String>,
        args: Array<JSONSerializable?> = emptyArray(),
        callback: ((String?, Any?) -> Unit)
    ) {
        val promiseId = UUID.randomUUID().toString()
        promises[promiseId] = callback

        val segmentArray = JSONArray(identifierSegments)
        val segmentString = segmentArray.toString()

        val jsonArray = JSONArray()
        args.forEachIndexed { _, jsonSerializable ->
            val asPrimitive = jsonSerializable as? PrimitiveJSONSerializable
            if (asPrimitive != null) {
                jsonArray.put(asPrimitive.value)
            } else {
                jsonArray.put(if (jsonSerializable == null) JSONObject.NULL
                else JSONObject(jsonSerializable.stringify()))
            }
        }
        val jsonString = jsonArray.toString()
        val script = """
        {
            let idSegments = $segmentString;
            let args = $jsonString;
            let promise = undefined;
            try {
                let fn = idSegments.reduce((state, key) => {
                    return state[key];
                }, window);
                promise = Promise.resolve(fn(...args));
            } catch (error) {
                promise = Promise.reject(error);
            }
            promise.then((value) => {
                _nimbus.resolvePromise("$promiseId", JSON.stringify({value: value}));
            }).catch((err) => {
                _nimbus.rejectPromise("$promiseId", err.toString());
            });
        }
        null;
        """.trimIndent()

        bridgeWebView?.handler?.post {
            bridgeWebView?.evaluateJavascript(script, null)
        }
    }

    fun invoke(
        functionName: String,
        args: Array<JSONSerializable?> = emptyArray(),
        callback: ((String?, Any?) -> Unit)
    ) {
        invokeInternal(functionName.split('.').toTypedArray(), args, callback)
    }

    @Suppress("unused")
    @JavascriptInterface
    fun resolvePromise(promiseId: String, json: String?) {
        var value: Any? = null
        json?.let {
            value = JSONObject(it).get("value")
        }
        val promise = promises.remove(promiseId)
        promise?.let {
            it(null, value)
        }
    }

    @Suppress("unused")
    @JavascriptInterface
    fun rejectPromise(promiseId: String, error: String) {
        val promise = promises.remove(promiseId)
        promise?.let {
            it(error, null)
        }
    }

    @Suppress("unused")
    @JavascriptInterface
    fun pageUnloaded() {
        val canceledPromises = ConcurrentHashMap(promises)
        promises.clear()
        for (promise in canceledPromises.values) {
            promise("ERROR_PAGE_UNLOADED", null)
        }
    }

    private val promises: ConcurrentHashMap<String, Function2<String?, Any?, Unit>> = ConcurrentHashMap()

    /**
     * Creates and returns a Callback object that can be passed as an argument to
     * a subsequent JavascriptInterface bound method.
     */
    @Suppress("unused")
    @JavascriptInterface
    fun makeCallback(callbackId: String): Callback? {
        return bridgeWebView?.let { return Callback(it, callbackId) }
    }

    /**
     * Return the names of all connected extensions so they can be processed by the
     * JavaScript runtime code.
     */
    @Suppress("unused")
    @JavascriptInterface
    fun nativePluginNames(): String {
        val names = binders.map { it.getPluginName() }
        val result = JSONArray(names)
        return result.toString()
    }

    private fun initialize(webView: WebView, binders: Collection<Binder>) {
        binders.forEach { binder ->

            // customize web view if needed
            binder.getPlugin().customize(webView, this)

            // bind web view to binder
            binder.setWebView(webView)

            // add the javascript interface for the binder
            val extensionName = binder.getPluginName()
            webView.addJavascriptInterface(binder, "_$extensionName")
        }
    }

    protected fun finalize() {
        promises.values.forEach { it.invoke("Canceled", null) }
        promises.clear()
    }

    private fun cleanup(webView: WebView, binders: Collection<Binder>) {
        binders.forEach { binder ->

            // cleanup web view if needed
            binder.getPlugin().cleanup(webView, this)

            // unbind web view from binder
            binder.setWebView(null)

            // remove the javascript interface for the binder
            val extensionName = binder.getPluginName()
            webView.removeJavascriptInterface("_$extensionName")
        }
    }
}

File: platforms/apple/Sources/Nimbus/Bridge.swift
//
// Copyright (c) 2019, Salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
//

import Foundation
import WebKit

public class Bridge: NSObject, JSEvaluating {
    public func addPlugin<T: Plugin>(_ plugin: T) {
        plugins.append(plugin)
    }

    public func attach(to webView: WKWebView) {
        guard self.webView == nil else {
            return
        }

        self.webView = webView
        let configuration = webView.configuration
        configuration.userContentController.add(self, name: "_nimbus")
        configuration.preferences.javaScriptEnabled = true
        #if DEBUG
            configuration.preferences.setValue(true, forKey: "developerExtrasEnabled")
        #endif

        for plugin in plugins {
            let connection = WebViewConnection(from: webView, bridge: self, as: plugin.namespace)
            plugin.bind(to: connection)
        }
    }

    /**
     Invokes a Promise-returning Javascript function and call the specified
     promiseCompletion when that Promise resolves or rejects.
     */
    func invoke<R>( // swiftlint:disable:this function_body_length
        _ identifierSegments: [String],
        with args: [Encodable],
        callback: @escaping (Error?, R?) -> Void
    ) {
        let promiseId = UUID().uuidString
        promisesQueue.sync {
            self.promises[promiseId] = { error, value in
                if error != nil {
                    callback(error, nil)
                } else {
                    if R.self == Void.self {
                        callback(nil, nil)
                    } else if let result = value as? R {
                        callback(nil, result)
                    } else {
                        callback(
                            PromiseError.message(
                                "Could not convert \(String(describing: value)) to \(R.self)"
                            ),
                            nil
                        )
                    }
                }
            }
        }

        let idSegmentString: String
        let argString: String
        do {
            let data = try JSONEncoder().encode(args.map(EncodableValue.value))
            argString = String(data: data, encoding: .utf8)!

            let idData = try JSONEncoder().encode(identifierSegments)
            idSegmentString = String(data: idData, encoding: .utf8)!
        } catch {
            return callback(error, nil)
        }

        let script = """
        {
            let idSegments = \(idSegmentString);
            let rawArgs = \(argString);
            let args = rawArgs.map(a => a.v);
            let promise = undefined;
            try {
                let fn = idSegments.reduce((state, key) => {
                    return state[key];
                }, window);
                promise = Promise.resolve(fn(...args));
            } catch (error) {
                promise = Promise.reject(error);
            }
            promise.then((value) => {
                webkit.messageHandlers._nimbus.postMessage({
                    method: "resolvePromise",
                    promiseId: "\(promiseId)",
                    value: value
                });
            }).catch((err) => {
                webkit.messageHandlers._nimbus.postMessage({
                    method: "rejectPromise",
                    promiseId: "\(promiseId)",
                    value: err.toString()
                });
            });
        }
        null;
        """

        webView?.evaluateJavaScript(script) { _, error in
            if let error = error {
                var callback: PromiseCallback?
                self.promisesQueue.sync {
                    callback = self.promises.removeValue(forKey: promiseId)
                }
                callback?(error, nil)
            }
        }
    }

    /**
     Invokes a Promise-returning Javascript function and call the specified
     promiseCompletion when that Promise resolves or rejects.
     */
    public func evaluate<R: Decodable>(
        _ identifierPath: String,
        with args: [Encodable],
        callback: @escaping (Error?, R?) -> Void
    ) {
        let identifierSegments = identifierPath.split(separator: ".").map(String.init)
        invoke(identifierSegments, with: args, callback: callback)
    }

    var plugins: [Plugin] = []
    private let promisesQueue = DispatchQueue(label: "Nimbus.promisesQueue")
    typealias PromiseCallback = (Error?, Any?) -> Void
    private var promises: [String: PromiseCallback] = [:]
    weak var webView: WKWebView?
}

extension Bridge: WKScriptMessageHandler {
    public func userContentController(_: WKUserContentController, didReceive message: WKScriptMessage) {
        guard let body = message.body as? [String: Any] else { return }
        guard let method = body["method"] as? String else { return }

        switch method {
        case "resolvePromise":
            guard let promiseId = body["promiseId"] as? String else {
                    return
            }
            var callback: PromiseCallback?
            self.promisesQueue.sync {
                callback = self.promises.removeValue(forKey: promiseId)
            }
            callback?(nil, body["value"])

        case "rejectPromise":
            guard let promiseId = body["promiseId"] as? String,
                let value = body["value"] as? String else {
                    return
            }
            var callback: PromiseCallback?
            self.promisesQueue.sync {
                callback = self.promises.removeValue(forKey: promiseId)
            }
            callback?(PromiseError.message(value), nil)

        case "pageUnloaded":
            var callbacks: [String: PromiseCallback] = [:]
            self.promisesQueue.sync {
                callbacks = self.promises
                self.promises = [:]
            }
            callbacks.values.forEach { callback in callback(PromiseError.pageUnloaded, nil) }

        default:
            break
        }
    }
}

public enum PromiseError: Error, Equatable {
    case pageUnloaded
    case message(_ message: String)
}
